include/xtensor/xreducer.hpp

```diff
@@ -37,20 +37,231 @@ namespace xt
+            if (*it == last_ax - 1)
+            {
+                last_ax = *it;
+                outer_loop_size *= e.shape()[last_ax];
+            }
+        }
+
+        idx = 0;
+        for (std::size_t i = 0; i < e.dimension(); ++i)
+        {
+            if (std::find(axes.begin(), axes.end(), i) == axes.end())
+            {
+                // i not in axes!
+                iter_strides[i] = result.strides()[idx];
+                ++idx;
+            }
+        }
+
+        // TODO remove all entries where iter_shape == 0
+        iter_shape.erase(iter_shape.begin() + last_ax, iter_shape.end());
+        iter_strides.erase(iter_strides.begin() + last_ax, iter_strides.end());
+
+        for (std::size_t i = 0; i < iter_shape.size(); ++i)
+        {
+            if (iter_shape[i] == 0)
+            {
+                iter_shape.erase(iter_shape.begin() + i);
+                iter_strides.erase(iter_strides.begin() + i);
+                ++i;
+            }
+        }
+
+        xindex temp_idx(iter_shape.size());
+        auto next_idx = [&iter_shape, &iter_strides, &temp_idx]()
+        {
+            int i = iter_shape.size() - 1;
+            for (; i >= 0; --i)
+            {
+                if (ptrdiff_t(temp_idx[(std::size_t) i]) >= ptrdiff_t(iter_shape[(std::size_t) i]) - 1)
+                {
+                    temp_idx[(std::size_t) i] = 0;
+                }
+                else
+                {
+                    temp_idx[(std::size_t) i]++;
+                    break;
+                }
+            }
+            return std::make_pair(i == -1,
+                                  std::inner_product(temp_idx.begin(), temp_idx.end(),
+                                                     iter_strides.begin(), ptrdiff_t(0)));
+        };
+
+        auto begin = e.raw_data();
+        auto out = result.raw_data();
+        auto out_begin = result.raw_data();
+
+        ptrdiff_t next_stride = 0;
+
+        std::pair<bool, ptrdiff_t> idx_res(false, 0);
+
+        // Remark: eventually some modifications here to make conditions faster where merge + accumulate is the
+        // same function (e.g. check std::is_same<decltype(merge_fct), decltype(acc_fct)>::value) ...
+
+        auto merge_border = out;
+        bool merge = false;
+        // Decide if going about it row-wise or col-wise
+        if (inner_stride == 1)
+        {
+            while(idx_res.first != true)
+            {
+                // for unknown reasons it's much faster to use a temporary variable and
+                // std::accumulate here -- probably some cache behavior
+                result_type tmp;
+                tmp = init_fct(*begin);
+                tmp = std::accumulate(begin + 1, begin + outer_loop_size, tmp, acc_fct);
+
+                if (merge)
+                {
+                    *out = merge_fct(*out, tmp);
+                }
+                else
+                {
+                    *out = tmp;
+                }
```

Comment:
maybe replace with `*out = merge ? merge_fct(*out, tmp) : tmp;`
The code is long enough so that using a ternary conditional operator is justified.

Reply:
:+1: